File: apps/server/app/services/receipt/sample.py
# -*- coding: utf-8 -*-
"""NAMUNA chek — sozlama oldindan ko'rinishi va sinov chop etishi uchun (Phase 5F).

⚠️  BAZAGA HECH NARSA YOZILMAYDI: faqat shablon/do'kon ma'lumoti O'QILADI, qolgani
    sintetik. `test: true`, raqam «TEST», `doc.id` yo'q — bu hujjat hech qayerda
    mavjud emas va hech qanday hisobotga tushmaydi (`tests/test_receipt_no_writes.py`
    barcha jadvallar izi bilan isbotlaydi).
⚠️  Summalar HAQIQIY chek qoidasi bilan hisoblanadi (butun so'mga yaxlitlash, to'lovlar
    yig'indisi = jami): namuna renderer'ning arifmetika tekshiruvini ham sinaydi.
"""
from __future__ import annotations

from datetime import datetime, timezone
from decimal import ROUND_CEILING, ROUND_HALF_UP, Decimal

from sqlalchemy.orm import Session

from app.models.org import Branch, Company
from app.services.receipt.dto import (_logo_ref, assemble, branch_tz, iso_utc, local_str, money,
                                      qty, totals_block)
from app.services.receipt.settings import resolve_for_branch, resolve_store

KINDS = ("sale", "mixed", "return", "long")
SAMPLE_NUMBER = "TEST"
# Shtrix-kod/QR namunasi: aniq TEST belgisi — skanerlansa hech qanday chekka mos kelmaydi.
SAMPLE_UID = "TEST000000"

# (nom, miqdor, narx, birlik, og'irlikli, qator chegirmasi)
_SALE = [
    ("Non «Buxanka»", "2", "3500", "dona", False, "0"),
    ("Sut 3,2% 1 l", "1", "12000", "dona", False, "0"),
    ("Pomidor (issiqxona)", "0.347", "18900", "kg", True, "0"),
    ("Choy «Ahmad» 100 g", "1", "24500", "dona", False, "0"),
]
_MIXED = [
    ("Guruch «Lazer» 1 kg", "3", "16000", "dona", False, "1500"),
    ("Qo'y go'shti (lahm)", "1.254", "98000", "kg", True, "0"),
    ("Молоко «Простоквашино» 2,5%", "2", "14500", "dona", False, "0"),
    ("Kungaboqar yog'i 1 l", "1", "21000", "dona", False, "1000"),
]
_RETURN = [
    ("Sut 3,2% 1 l", "1", "12000", "dona", False, "0"),
    ("Olma «Semerenko»", "0.815", "16000", "kg", True, "0"),
]
_LONG_NAMES = [
    ("Молоко пастеризованное «Простоквашино» 3,2% 930 мл", "dona", False),
    ("Көмөч нан «Токмок» — жаңы бышырылган, 400 г", "dona", False),
    ("Балмуздак «Үлгү» ванилдүү, 80 г", "dona", False),
    ("Oʻzbekiston choyi «Gʻallaorol» koʻk choy, 100 g", "dona", False),
    ("Ўзбекистон шакари оқ, қадоқланган 1 кг — ҳалол", "dona", False),
    ("Qoʻy goʻshti (lahm) — yangi soʻyilgan, 1-nav", "kg", True),
    ("Superuzunmahsulotnomibo'shliqsiz0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ", "dona", False),
    ("Бадыраң (тепличный), сорт «Өрнөк»", "kg", True),
]


def _lines(spec) -> tuple[list, Decimal, Decimal]:
    out, subtotal, disc = [], Decimal("0"), Decimal("0")
    for name, q, price, unit, weighted, d in spec:
        qd, pd, dd = Decimal(q), Decimal(price), Decimal(d)
        gross = (qd * pd).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
        out.append({"name": name, "qty": qty(qd), "unit": unit, "weighted": weighted,
                    "unit_price": money(pd), "gross": money(gross), "discount": money(dd),
                    "total": money(gross - dd)})
        subtotal += gross
        disc += dd
    return out, subtotal, disc


def _long_spec() -> list:
    spec = []
    for i in range(120):
        name, unit, weighted = _LONG_NAMES[i % len(_LONG_NAMES)]
        if weighted:
            q = "0.001" if i % 16 == 5 else f"{(i % 7) + 1}.{(i * 37) % 1000:03d}"
        else:
            q = str((i % 4) + 1)
        price = str(1000 + (i * 7919) % 250_000)
        disc = str((i % 5) * 500) if i % 5 == 0 and i else "0"
        spec.append((f"{i + 1:03d} {name}", q, price, unit, weighted, disc))
    # Katta summa: chek kengligi 9–10 xonali jamini ham sig'dirishini tekshiradi.
    spec.append(("Televizor 75\" QLED (kafolat 3 yil)", "39", "25200000", "dona", False, "345000"))
    return spec


def build_sample(db: Session, emp, branch: Branch, kind: str, *,
                 now: datetime | None = None) -> dict:
    company = db.get(Company, emp.company_id)
    eff, _, _ = resolve_for_branch(db, emp.company_id, branch)
    store = resolve_store(db, company, branch, eff)
    tzname, tz = branch_tz(branch)
    now = now or datetime.now(timezone.utc)
    currency = (company.currency if company else None) or "UZS"
    is_return = kind == "return"
    spec = {"sale": _SALE, "mixed": _MIXED, "return": _RETURN, "long": None}[kind] or _long_spec()
    lines, subtotal, line_disc = _lines(spec)
    doc_disc = Decimal("2000") if kind in ("mixed", "long") else Decimal("0")
    total = (subtotal - line_disc - doc_disc).quantize(Decimal("1"), rounding=ROUND_HALF_UP)

    payments: list = []
    refund = original = None
    if is_return:
        # Qaytarish: qator summasi = jami (chegirma yo'q), yaxlitlash — ochiq qator.
        for ln in lines:
            ln["discount"] = "0.00"
            ln["total"] = ln["gross"]
        line_disc = Decimal("0")
        refund = {"method": "cash", "amount": money(total)}
        original = {"id": None, "number": SAMPLE_NUMBER, "uid": None,
                    "issued_at_local": local_str(now, tz)}
    elif kind == "sale":
        given = (total / Decimal("10000")).to_integral_value(rounding=ROUND_CEILING) * 10000
        payments = [{"method": "cash", "amount": money(total), "given": money(given),
                     "change": money(given - total)}]
    else:
        cash = (total * Decimal("0.2")).quantize(Decimal("1"), rounding=ROUND_HALF_UP)
        qr = (total * Decimal("0.1")).quantize(Decimal("1"), rounding=ROUND_HALF_UP)
        card = total - cash - qr
        given = (cash / Decimal("1000")).to_integral_value(rounding=ROUND_CEILING) * Decimal("1000")
        payments = [{"method": "cash", "amount": money(cash), "given": money(given),
                     "change": money(given - cash)},
                    {"method": "card", "amount": money(card), "given": None, "change": None},
                    {"method": "qr", "amount": money(qr), "given": None, "change": None}]

    doc = {"id": None, "number": SAMPLE_NUMBER, "uid": None if is_return else SAMPLE_UID,
           "issued_at": iso_utc(now), "issued_at_local": local_str(now, tz), "tz": tzname,
           "is_offline": False, "status": "completed"}
    actor = {"cashier": emp.full_name, "till_code": "TEST", "terminal": None}
    customer = {"name": "TEST"} if eff["show_customer"] else None
    return assemble(
        kind="RETURN" if is_return else "SALE", test=True, doc=doc, store=store, actor=actor,
        customer=customer, lines=lines,
        totals=totals_block(currency, subtotal, line_disc, Decimal("0") if is_return else doc_disc,
                            total),
        payments=payments, refund=refund, original=original, eff=eff,
        logo=_logo_ref(db, emp.company_id, branch.id, eff))

File: packages/shared/src/receipt/sample.ts
// Oflayn namuna chek (TEST) — server `/receipt/sample` javob bermaganda Manager oldindan ko'rishi va
// test chop etish uchun. HAMMA summalar o'nlik yordamchilar bilan hisoblanadi: namuna ham server
// invariantlarini saqlaydi (oraliq − chegirmalar + yaxlitlash = JAMI, Σ to'lovlar = JAMI).
//
// Deterministik: sana o'zgarmas (TEST banneri bilan chiqadi); kerak bo'lsa chaqiruvchi
// `dto.doc.issued_at_local` ni almashtiradi. QR matritsasi bu yerda hisoblanmaydi (QR kodlovchi yo'q) —
// `qr: null`; shtrix-kod esa CODE128-B bilan (code128.ts).
import type { PayMethod, ReceiptDTO, ReceiptLine, ReceiptTemplate } from "./types";
import { BUILTIN_TEMPLATE } from "./types";
import { decMul, decRound, decSub, decSum, decCmp } from "./format";
import { normalizeTemplate } from "./layout";
import { code128Modules } from "./code128";

export type SampleKind = "sale" | "mixed" | "return" | "long";

const ISSUED_AT = "2026-09-19T09:32:11+00:00";
const ISSUED_AT_LOCAL = "19.09.2026 14:32";
const SAMPLE_UID = "0000000000";

function line(name: string, qty: string, unit: string | null, weighted: boolean, price: string, discount = "0.00"): ReceiptLine {
  const q = decRound(qty, 3);
  const p = decRound(price, 2);
  const gross = decMul(q, p, 2);
  const d = decRound(discount, 2);
  return { name, qty: q, unit, weighted, unit_price: p, gross, discount: d, total: decSub(gross, d, 2) };
}

const SALE_LINES = (): ReceiptLine[] => [
  line("Non \"Toshkent\" 500 g", "2", "dona", false, "4000"),
  line("Pomidor", "0.352", "kg", true, "12000"),
  line("Молоко 3,2% 1 л", "1", "dona", false, "11500", "500"),
];

const MIXED_LINES = (): ReceiptLine[] => [
  ...SALE_LINES(),
  line("Olma Golden", "1.235", "kg", true, "18990"),
  line("Шоколад молочный 90 г", "3", "dona", false, "8750.50"),
];

const RETURN_LINES = (): ReceiptLine[] => [
  line("Молоко 3,2% 1 л", "1", "dona", false, "11000"),
  line("Pomidor", "0.352", "kg", true, "12000"),
];

// Uzun nomlar: rus, qirg'iz (ң ө ү), o'zbek lotin (oʻ gʻ — U+02BB), o'zbek kirill, bo'shliqsiz uzun so'z.
const LONG_NAMES = [
  "Молоко ультрапастеризованное «Простоквашино» 3,2% 950 мл",
  "Кымыз табигый, жаңы саалган 1 л (Суусамыр өрөөнү)",
  "Oʻzbekiston paxta yogʻi tozalangan 1 l (Gʻallaorol)",
  "Ўзбекистон гилоси, янги узилган (Фарғона водийси)",
  "Колбаса варёная «Докторская» ГОСТ высший сорт 500 г",
  "СуперДлинноеНазваниеТовараБезПробеловКотороеНеПомещаетсяВСтроку",
  "Өрүк кургатылган Баткен, жогорку сорт, үй шартында",
  "Qoʻy goʻshti yangi soʻyilgan, suyaksiz",
];
const LONG_PRICES = ["4000", "12990.50", "250000", "1500", "86400", "3999.99", "17250"];

function longLines(): ReceiptLine[] {
  const out: ReceiptLine[] = [];
  for (let i = 0; i < 120; i++) {
    const name = `${LONG_NAMES[i % LONG_NAMES.length]} #${i + 1}`;
    const price = LONG_PRICES[i % LONG_PRICES.length];
    const weighted = i % 5 === 1;
    const qty = weighted ? ["0.001", "1.235", "0.352", "12.500"][Math.floor(i / 5) % 4] : String((i % 4) + 1);
    let l = line(name, qty, weighted ? "kg" : "dona", weighted, price);
    if (i % 7 === 3) l = line(name, qty, weighted ? "kg" : "dona", weighted, price, decMul(l.gross, "0.05", 2));
    out.push(l);
  }
  // Katta summa: 9 xonali jami (987 654 321 atrofi) — minglik ajratgich va kenglik sinovi.
  out.push(line("Sanoat muzlatkichi kompleksi (ombor uchun), oʻrnatish bilan", "8", "dona", false, "98765432.10"));
  return out;
}

function totalsOf(lines: ReceiptLine[], docDiscount: string) {
  const subtotal = decSum(lines.map((l) => l.gross), 2);
  const line_discount = decSum(lines.map((l) => l.discount), 2);
  const doc_discount = decRound(docDiscount, 2);
  const net = decSub(decSub(subtotal, line_discount, 2), doc_discount, 2);
  const total = decRound(decRound(net, 0), 2); // server kabi butun so'mga ROUND_HALF_UP
  return { currency: "UZS", subtotal, line_discount, doc_discount, rounding: decSub(total, net, 2), total };
}

const pay = (method: PayMethod, amount: string, given: string | null = null, change: string | null = null) => ({
  method, amount: decRound(amount, 2), given: given === null ? null : decRound(given, 2),
  change: change === null ? null : decRound(change, 2),
});

export function sampleReceipt(
  kind: SampleKind,
  store?: Partial<ReceiptDTO["store"]>,
  template?: Partial<ReceiptTemplate>,
): ReceiptDTO {
  const T = normalizeTemplate({ ...BUILTIN_TEMPLATE, ...(template ?? {}) });
  const k: SampleKind = kind === "mixed" || kind === "return" || kind === "long" ? kind : "sale";
  const isReturn = k === "return";
  const lines = k === "sale" ? SALE_LINES() : k === "mixed" ? MIXED_LINES() : k === "return" ? RETURN_LINES() : longLines();
  const totals = totalsOf(lines, k === "long" ? "1000" : "0");

  let payments: ReceiptDTO["payments"] = [];
  if (k === "sale") {
    // Naqd: xaridor 25 000 berdi (jami undan oshsa — aniq summa), qaytim = berilgan − jami.
    const g = decCmp(totals.total, "25000") <= 0 ? "25000.00" : totals.total;
    payments = [pay("cash", totals.total, g, decSub(g, totals.total, 2))];
  } else if (k === "mixed" || k === "long") {
    const cash = k === "long" ? "150000.00" : "20000.00";
    const qr = k === "long" ? "250000.00" : "15000.00";
    payments = [pay("cash", cash), pay("qr", qr), pay("card", decSub(decSub(totals.total, cash, 2), qr, 2))];
  }

  const uid = isReturn ? null : SAMPLE_UID;
  const modules = uid && T.show_barcode ? code128Modules(uid) : null;
  return {
    schema: "binos.receipt.v1",
    kind: isReturn ? "RETURN" : "SALE",
    test: true,
    provisional: false,
    doc: {
      id: null, number: "TEST", uid, issued_at: ISSUED_AT, issued_at_local: ISSUED_AT_LOCAL,
      tz: "Asia/Tashkent", is_offline: false, status: "completed",
    },
    store: {
      name: store?.name ?? "SavdoOS",
      branch_name: store?.branch_name ?? null,
      address: store?.address ?? null,
      phone: store?.phone ?? null,
      stir: store?.stir ?? null,
    },
    actor: { cashier: "Dilnoza Karimova", till_code: "K-01", terminal: null },
    customer: T.show_customer ? { name: "Ali Valiyev" } : null,
    lines,
    totals,
    payments,
    refund: isReturn ? { method: "cash", amount: totals.total } : null,
    original: isReturn ? { id: null, number: "TEST", uid: null, issued_at_local: ISSUED_AT_LOCAL } : null,
    barcode: uid && modules ? { format: "CODE128", payload: uid, modules } : null,
    qr: null,
    logo: null,
    template: T,
  };
}
